src/main/scala/pipedsl/common/Utilities.scala

```diff
@@ -679,4 +697,66 @@ object Utilities {
   val is_handle_var :Id => Boolean =
     { id: Id => id.v.startsWith(lock_handle_prefix) }
 
+  val generic_type_prefix = "__SUPERSPECIALGENERIC_"
+  def is_generic(t :Any) :Boolean = t match {
+    case TNamedType(name) => name.v.startsWith(generic_type_prefix)
+    case TSizedInt(l, _) => is_generic(l)
+    case TBitWidthVar(name) => name.v.startsWith(generic_type_prefix)
+    case _:Type => false
+    case id:Id => id.v.startsWith(generic_type_prefix)
+    case s:String => s.startsWith(generic_type_prefix)
+  }
+
+  val not_gen_pref = "__NOT_GEN"
+  def degenerify(t :Type) :Type =
+    {
+      def _degenerify(t :Type) :Type = t match
+      {
+        case TSignVar(nm) if (is_generic(nm)) =>
+          TSignVar(Id(not_gen_pref + nm.v).copyMeta(nm))
+        case TSizedInt(len, sign) =>
+          TSizedInt(degenerify(len).asInstanceOf[TBitWidth], degenerify(sign).asInstanceOf[TSignedNess])
+        case TFun(args, ret) =>
+          TFun(args.map(degenerify), degenerify(ret))
+        case TRecType(name, fields) =>
+          TRecType(name, fields.map((idtp) => (idtp._1, degenerify(idtp._2))))
+        case TModType(inputs, refs, retType, name) =>
+          TModType(inputs.map(degenerify), refs.map(degenerify), retType.map(degenerify), name)
+        case TReqHandle(tp, rtyp) => TReqHandle(degenerify(tp), rtyp)
+        case TNamedType(name) if is_generic(name) =>
+          TNamedType(Id(not_gen_pref + name.v).copyMeta(name))
+        case TMaybe(btyp) =>
+          TMaybe(degenerify(btyp))
+        case TBitWidthVar(nm) if is_generic(nm) =>
+          TBitWidthVar(Id(not_gen_pref + nm.v).copyMeta(nm))
+        case TObject(name, typParams, methods) =>
+          TObject(name, typParams.map(degenerify), methods.map(idtp => (idtp._1, (degenerify(idtp._2._1).asInstanceOf[TFun], idtp._2._2))))
+        case other => other
+      }
+      _degenerify(t).copyMeta(t)
+    }
+
+  def specialise(t :Type, s :List[Int]) :Type = t match {
```

Comment:
could use comments on these few functions (since they're library functions) - I'm not really sure what they're all for

Reply:
a valid concern. Done